### ISSUE TITLE
Feat request: entries soft delete

### ISSUE THREAD

**Issue by wohenbushuang** [meeting-room-booking-system/mrbs-code#73]
**Is your feature request related to a problem? Please describe.**
Someone told me that their booking was missing and show me the booking successful screenshot for proof. When I need to check the backend log, I found that there is no old entry log left in mysql database.

**Describe the solution you'd like**
When deleting such as repeats, entries, use soft delete to keep the booking log for audit backtracing.

**Describe alternatives you've considered**
To keep a user operational log would be better

**Comment by campbell-m** [comment]
One thing you can do in the meantime is configure MRBS so that all deletions generate an email notification to the admin.  If you set up a special email account for the admin (eg a gmail account) you will effectively have a log of all deletions, including the name of the person who deleted the entry.  The log could also contain additions and edits, depending on how you configure the mail settings.

Note that deletions using the bulk deletion button on the Report page don't generate an email, but if that's a concern you could always disable that feature in the config file.

**Comment by MarkoD** [comment]
@campbell-m We would also like to implement this feature in order to track "missed bookings" and "no-shows". Is there a way to perhaps alter MRBS - instead of sending a "delete entry" e-mail notification, to move the current entry to a separate (new) table, including a user-chosen text detail (such as "no-show" or "cancelled after grace period")?

**Comment by campbell-m** [comment]
Since Version 1.12.0, MRBS has included [the Monolog logger](https://github.com/Seldaek/monolog).  Maybe one way of logging events would be to use Monolog to log events of interest to a database (or a file, or email)?

**Comment by MarkoD** [comment]
@campbell-m  Thanks for the heads up. However, if we would like to ask in the UI (when the user clicks od "Delete booking") for the reason of the cancellation (i.e. "no show" or "late cancellation" or etc.) and store the string besides the deleted booking, what would be the simplest way to achieve that?

**Comment by campbell-m** [comment]
I don't know what the best way to do it would be.  The simplest would be to add a select element next to the Delete button, which is required to be filled in with the reason.   But from a UI point of view it might be better to direct the user to another page after they've pressed the Delete button.  That page would ask them for the reason for cancellation and also to confirm the cancellation.

**Comment by wohenbushuang** [comment]
I wonder whether adding a `deleted_at` like column in the table as well as relavant backend logic changes are acceptable? If you agree with so, maybe someone can implement this and create a PR.

**Comment by campbell-m** [comment]
>  I wonder whether adding a deleted_at like column in the table

Which table?  If you mean the mrbs_entry table, then I don't think that will help as the row will immediately be deleted.

**Comment by MarkoD** [comment]
@wohenbushuang That would require to also modify all controllers and models in MRBS to ignore entries with `deleted_at != null`, which I guess is a lot of work... Perhaps an easier approach would be to create a separate table into which you would copy entries just before they are deleted from `mrsb_entry` - that way you only need to modify the delete action/code in MRBS (`edit_entry.php` and `view_entry.php`).

**Comment by campbell-m** [comment]
Yes, a separate table would be much better and you coud use the Monolog logger to write to it.  You could also use the room, area and user descriptions at the time the entry was deleted, instead of ids, which may change by the time you come to read the log.

**Comment by wohenbushuang** [comment]
> Which table? If you mean the mrbs_entry table, then I don't think that will help as the row will immediately be deleted.

Yep I mean the `mrbs_entry` table, and "relevant backend logic changes" means

> modify all controllers and models in MRBS to ignore entries with `deleted_at != null`

---

> which I guess is a lot of work...

So I think this need @campbell-m decision for the future of this project, or it might waste a lot of effort.

---

> create a separate table into which you would copy entries just before they are deleted

I agree that it is a quick feasible solution for our private instances (just like I have used email log as the temporal solution since I created this issue), but it seems not elegant for this project.

**Comment by MarkoD** [comment]
@wohenbushuang To sum up, you are asking @campbell-m whether the "sotf-deletion" of reservation entries is planned in the future for MRBS.

**Comment by campbell-m** [comment]
Wouldn't you also need to have a column "deleted_by", because presumably that's what people want to know when they find their booking's been deleted.  And @MarkoD would also need a "deleted_reason".

And would other people want to know similar information for users, areas and rooms?

**Comment by wohenbushuang** [comment]
> Wouldn't you also need to have a column "deleted_by", because presumably that's what people want to know when they find their booking's been deleted. And [@MarkoD](https://github.com/MarkoD) would also need a "deleted_reason".

How about reuse such as `modified_by`, `info_text` for simplicity?
- flag column `deleted_at IS NOT NULL` means the booking has been deleted.
  - `modified_by == created_by` means deleted by the booking creator.
  - `modified_by <> created_by` means deleted, i.e., rejected (thus also work for reject in approvals), by admin. Reasons can be seen in `info_text`.

> And would other people want to know similar information for users, areas and rooms?

I prefer these in a new `log` table. 😆

**Comment by MarkoD** [comment]
We've slightly modified the delete button action (in `edit_entry.php` and `view_entry.php`) to ask the user for the reason and store it, including a copy of the to-be-deleted entry in a separate table, so we're good. This approach is simple and does not require significant changes to MRBS.

If we wouldn't need an extra column (reason for deletion), than an even easier would be to just watch Monolog for delete entries.

Thank you for all insights, MRBS is a great piece of software!
